File: backtest/backtest.py
from typing import Dict, List
import ccxt
import pandas as pd
from datetime import datetime
from config import BacktestConfig
from strategy import PriceActionStrategy

class Backtester:
    def __init__(self, config: BacktestConfig):
        self.config = config
        self.strategy = PriceActionStrategy()
        self.exchange = ccxt.binance()
        self.results: Dict[str, Dict] = {}
        self.data_summary: Dict[str, Dict] = {}
        self.trade_history: Dict[str, List] = {}
        
    def fetch_data(self, symbol: str) -> pd.DataFrame:
        start_timestamp = self.exchange.parse8601(self.config.start_date)
        end_timestamp = self.exchange.parse8601(self.config.end_date)
        
        ohlcv = self.exchange.fetch_ohlcv(
            symbol,
            timeframe=self.config.timeframe,
            since=start_timestamp,
            limit=1000
        )
        
        df = pd.DataFrame(
            ohlcv,
            columns=['timestamp', 'open', 'high', 'low', 'close', 'volume']
        )
        df['timestamp'] = pd.to_datetime(df['timestamp'], unit='ms')
        
        # Store data summary
        self.data_summary[symbol] = {
            'start_date': df['timestamp'].min(),
            'end_date': df['timestamp'].max(),
            'total_candles': len(df),
        }
        
        return df.set_index('timestamp')
    
    def run(self) -> Dict[str, Dict]:
        total_profit = 0
        
        for coin in self.config.coins:
            df = self.fetch_data(coin)
            df = self.strategy.calculate_signals(df)
            
            balance = self.config.initial_balance
            position = 0
            trades = 0
            winning_trades = 0
            self.trade_history[coin] = []
            
            for i in range(1, len(df)):
                if df['signal'].iloc[i] == 1 and position == 0:  # Buy
                    entry_price = df['close'].iloc[i]
                    position = balance / entry_price
                    balance -= balance * self.config.commission
                    trades += 1
                    trade_entry = {
                        'type': 'BUY',
                        'entry_price': entry_price,
                        'entry_date': df.index[i],
                    }
                    
                elif df['signal'].iloc[i] == -1 and position > 0:  # Sell
                    exit_price = df['close'].iloc[i]
                    new_balance = position * exit_price
                    trade_profit = new_balance - balance
                    balance = new_balance - (new_balance * self.config.commission)
                    
                    if trade_profit > 0:
                        winning_trades += 1
                    
                    self.trade_history[coin].append({
                        **trade_entry,
                        'exit_price': exit_price,
                        'exit_date': df.index[i],
                        'profit': trade_profit,
                    })
                    
                    position = 0
            
            # Close any remaining position
            if position > 0:
                exit_price = df['close'].iloc[-1]
                new_balance = position * exit_price
                trade_profit = new_balance - balance
                balance = new_balance - (new_balance * self.config.commission)
                
                self.trade_history[coin].append({
                    **trade_entry,
                    'exit_price': exit_price,
                    'profit': trade_profit,
                    'date': df.index[-1]
                })
            
            profit = balance - self.config.initial_balance
            profit_percentage = (profit / self.config.initial_balance) * 100
            win_rate = (winning_trades / trades * 100) if trades > 0 else 0
            avg_profit = profit / trades if trades > 0 else 0
            
            self.results[coin] = {
                'trades': trades,
                'profit': profit,
                'profit_percentage': profit_percentage,
                'win_rate': win_rate,
                'avg_profit_per_trade': avg_profit
            }
            
            total_profit += profit
        
        self.results['total'] = {'total_profit': total_profit}
        return self.results

File: backtest/ts/backtest.ts
import { CCXT } from 'ccxt';
import { DataFrame } from 'pandas-js';
import { BacktestConfig } from './config';
import { PriceActionStrategy } from './strategy';
import { 
    OHLCV, 
    TradeEntry, 
    DataSummary, 
    Results, 
    TradeStats 
} from './types';

export class Backtester {
    private readonly config: BacktestConfig;
    private readonly strategy: PriceActionStrategy;
    private readonly exchange: CCXT;
    private results: Results = {};
    private dataSummary: Record<string, DataSummary> = {};
    private tradeHistory: Record<string, TradeEntry[]> = {};

    constructor(config: BacktestConfig) {
        this.config = config;
        this.strategy = new PriceActionStrategy();
        this.exchange = new CCXT.binance();
    }

    private async fetchData(symbol: string): Promise<DataFrame> {
        const startTimestamp = this.exchange.parse8601(this.config.startDate);
        const endTimestamp = this.exchange.parse8601(this.config.endDate);
        
        const ohlcv = await this.exchange.fetchOHLCV(
            symbol,
            this.config.timeframe,
            startTimestamp,
            1000
        );
        
        const df = new DataFrame(ohlcv as OHLCV[]);
        
        this.dataSummary[symbol] = {
            startDate: df.get('timestamp').min(),
            endDate: df.get('timestamp').max(),
            totalCandles: df.length
        };
        
        return df;
    }

    public async run(): Promise<Results> {
        let totalProfit = 0;

        for (const coin of this.config.coins) {
            const df = await this.fetchData(coin);
            const signalDf = this.strategy.calculateSignals(df);
            
            let balance = this.config.initialBalance;
            let position = 0;
            let trades = 0;
            let winningTrades = 0;
            this.tradeHistory[coin] = [];
            
            for (let i = 1; i < signalDf.length; i++) {
                const row = signalDf.iloc(i);
                
                if (row.get('signal') === 1 && position === 0) {
                    const entryPrice = row.get('close');
                    position = balance / entryPrice;
                    balance -= balance * this.config.commission;
                    trades++;
                    
                    const tradeEntry: Partial<TradeEntry> = {
                        type: 'BUY',
                        entryPrice,
                        entryDate: row.get('timestamp')
                    };
                    
                    this.processTradeExit(
                        coin,
                        tradeEntry,
                        row,
                        position,
                        balance,
                        winningTrades
                    );
                }
            }
            
            this.calculateResults(
                coin,
                balance,
                trades,
                winningTrades,
                totalProfit
            );
        }
        
        this.results['total'] = { totalProfit };
        return this.results;
    }

    private processTradeExit(
        coin: string,
        tradeEntry: Partial<TradeEntry>,
        row: Series,
        position: number,
        balance: number,
        winningTrades: number
    ): void {
        const exitPrice = row.get('close');
        const newBalance = position * exitPrice;
        const tradeProfit = newBalance - balance;
        
        if (tradeProfit > 0) winningTrades++;
        
        this.tradeHistory[coin].push({
            ...tradeEntry as TradeEntry,
            exitPrice,
            exitDate: row.get('timestamp'),
            profit: tradeProfit
        });
    }

    private calculateResults(
        coin: string,
        balance: number,
        trades: number,
        winningTrades: number,
        totalProfit: number
    ): void {
        const profit = balance - this.config.initialBalance;
        const profitPercentage = (profit / this.config.initialBalance) * 100;
        const winRate = trades > 0 ? (winningTrades / trades) * 100 : 0;
        const avgProfit = trades > 0 ? profit / trades : 0;
        
        this.results[coin] = {
            trades,
            profit,
            profitPercentage,
            winRate,
            avgProfitPerTrade: avgProfit
        };
        
        totalProfit += profit;
    }
}